lib/pengine/utils.c

```diff
@@ -690,18 +690,35 @@ pe__update_recheck_time(time_t recheck, pcmk_scheduler_t *scheduler,
  * \param[in]     xml_obj       XML element containing blocks of nvpair elements
  * \param[in]     set_name      If not NULL, only use blocks of this element
  * \param[in]     rule_data     Matching parameters to use when unpacking
+ *                              (input argument only, but not const since it may
+ *                              be temporarily changed within this function)
  * \param[out]    hash          Where to store extracted name/value pairs
  * \param[in]     always_first  If not NULL, process block with this ID first
  * \param[in,out] scheduler     Scheduler data containing \p xml_obj
  */
 void
 pe__unpack_dataset_nvpairs(const xmlNode *xml_obj, const char *set_name,
-                           const pe_rule_eval_data_t *rule_data,
+                           pe_rule_eval_data_t *rule_data,
                            GHashTable *hash, const char *always_first,
                            pcmk_scheduler_t *scheduler)
 {
-    crm_time_t *next_change = crm_time_new_undefined();
+    crm_time_t *next_change = NULL;
+    GHashTable *orig_node_attrs = NULL;
 
+    CRM_CHECK((set_name != NULL) && (rule_data != NULL) && (hash != NULL)
+              && (scheduler != NULL), return);
+
+    if (xml_obj == NULL) {
+        return;
+    }
+
+    // Node attribute expressions are not allowed for meta-attributes
+    orig_node_attrs = rule_data->node_hash;
+    if (strcmp(set_name, PCMK_XE_META_ATTRIBUTES) == 0) {
```

Comment:
My preference would be to keep the argument const, and do a `CRM_CHECK` to ensure that `rule_data->node_hash == NULL` if `set_name` is `PCMK_XE_META_ATTRIBUTES`. That's the case for all possible callers as of this commit.

Reply:
makes sense